File: iir_regression.cc
/**\file
 * Implementation file for IIR regression routines.
 *
 * \author Rick Lan
 * \copyright See LICENSE for license.
 */

#include "fir_regression.h"
#include "iir_regression.h"
#include "iir.h"

#include <iostream>
#include <iomanip>
#include <string>
#include <vector>


void regress_iirbiquad(void)
{
	std::cout << __func__ << "..." << std::endl;
	std::string prefix = "regress_iirbiquad";

	std::vector<double> coeffs(7);
	coeffs[0] = 1.0;
	coeffs[1] = 0.0821;
	coeffs[2] = 1.0;
	coeffs[3] = 1.0;
	coeffs[4] = -1.2049;
	coeffs[5] = 0.5399;
	coeffs[6] = 0.1434; // gain
	write_to<double>(prefix + "_coeffs.txt", coeffs);

	IIRBiquad<double> iir(coeffs);
	const size_t n = 100;
	std::vector<double> out(n);
	for (size_t i = 0; i < n; i++) {
		if (i==0) {
			out[i] = iir.filter(1.0);
		} else {
			out[i] = iir.filter(0.0);
		}
	}
	write_to<double>(prefix + "_impulse_response.txt", out);
}

File: iir_regression.m
function []=iir_regression()
% IIR regression
% Reads output from executable and verify operation.
%
% Rick Lan
% See LICENSE for license.

regress_biquad();
end


function []=regress_biquad()
disp('regress_iirbiquad...');
c=load('regress_iirbiquad_coeffs.txt');
o=load('regress_iirbiquad_impulse_response.txt');
b=c(1:3)*c(7);
a=c(4:6);
ref=impz(b,a,100);
msg=sprintf('mse is %.12g', mse(ref,o));
disp(msg);
end


function [m] = mse(ref, cmp)
%MSE Compute mean squared error.
%   Y = MSE(REF, NOISY)
%   Input data structure supported: scalar, vector.
%   Input data type supported: real, complex.
%
%   Rick Lan
%   See LICENSE for the license.
%
error(nargchk(2, 2, nargin));
e = cmp - ref;
m = mean( e.*conj(e) );
end